src/Services/SearchService.php

```diff
@@ -306,24 +303,57 @@ private function shouldRedirectToProductDetailPage(array $productsIds, HttpReque
     /**
      * @return string|null
      */
-    private function getProductDetailUrl(int $productId)
+    private function getProductDetailUrl(Response $response)
     {
         /** @var ItemSearchService $itemSearchService */
         $itemSearchService = $this->getItemSearchService();
 
+        $productId = $response->getProductsIds()[0];
         $result = $itemSearchService->getResult(
-            $this->getSearchFactory([$productId])
+            $this->getSearchFactory()->hasItemId($productId)
         );
 
         if (!$result['success'] || empty($result['documents'][0])) {
             return null;
         }
 
-        $productData = $result['documents'][0]['data'];
+        $query = $response->getData(Response::DATA_QUERY)['query'];
 
+        $productData = $result['documents'][0]['data'];
         $urlPath = $productData['texts'][0]['urlPath'];
         $itemId = $productData['item']['id'];
+        $variationId = $this->getVariationIdForRedirect($query, $result['documents']);
+
+        return sprintf('/%s_%s_%s', $urlPath, $itemId, $variationId);
+    }
+
+    private function getVariationIdForRedirect(string $query, array $documents)
```

Comment:
1. The return type is not clear. Add one or add it in the docblock.
1. The method name does not suggest, that the main variation id will be returned, in case no appropriate variation can be found. Either add a docblock or update the method name.